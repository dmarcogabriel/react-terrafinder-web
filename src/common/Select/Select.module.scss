@import 'styles/';

.selectInput {
  position: relative;
  min-width: 10rem;

  .button {
    padding: $size-sm $size-sm;
    background: $white;
    border-radius: $size-md;
    width: 100%;
    cursor: pointer;

    .label {
      font-size: $size-sm;
      color: $orange;
    }

    .content {
      display: flex;
      flex-direction: row;
    }
  }
}

.selectOptionList {
  position: absolute;
  background: $white;

  .selectOption {
    width: 10rem;

    .selectOptionButton {
      display: flex;
      justify-content: center;
      width: 100%;
      padding: $size-sm 0;
      font-size: $size-xs;
      background: $white;
      color: $orange;
      cursor: pointer;
    }

    .selected {
      background: $orange;
      color: $white;
    }
  }
}

.selectArrowIcon {
  color: $orange;
}
