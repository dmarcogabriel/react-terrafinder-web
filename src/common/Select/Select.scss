@import 'styles/';

.selectInput {
  position: relative;
  min-width: 10rem;

  .searchInputButton {
    padding: $size-sm $size-sm;
    background: $white;
    width: 100%;
    cursor: pointer;

    .searchInputLabel {
      font-size: $size-sm;
      color: $orange;
    }

    .searchInputContent {
      display: flex;
      flex-direction: row;
    }
  }
}

.selectOptionList {
  min-width: 100%;
  position: absolute;
  background: $white;
  z-index: 50;
  box-shadow: 0 2px 10px 1px rgba(0, 0, 0, 0.5);
  padding: 1rem 0;
  display: flex;
  align-items: center;
  justify-content: center;

  .selectOptionButton {
    font-size: 1rem;
    background: $white;
    color: $orange;
    cursor: pointer;
  }

  .selectedOption {
    background: $orange;
    color: $white;
  }
}

.selectArrowIcon {
  color: $orange;
}
