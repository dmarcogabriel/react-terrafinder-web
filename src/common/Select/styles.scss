@import '../../styles/';

.selectInput {
  position: relative;
  width: 10rem;

  .selectInput--button {
    padding: $size-sm $size-xs;
    background: $white;
    border-radius: $size-md;
    cursor: pointer;

    .selectInput--button--label {
      font-size: $size-xs;
      color: $orange;
    }

    .selectInput--button--content {
      display: flex;
      flex-direction: row;
    }
  }
}

.selectOptionList {
  position: absolute;
  background: $white;

  .selectOptionList--option {
    width: 10rem;

    .selectOptionList--optionButton {
      display: flex;
      justify-content: center;
      width: 100%;
      padding: $size-sm 0;
      font-size: $size-xs;
      background: $white;
      color: $orange;
      cursor: pointer;
    }

    .selectOptionList--optionButton-selected {
      background: $orange;
      color: $white;
    }
  }
}
