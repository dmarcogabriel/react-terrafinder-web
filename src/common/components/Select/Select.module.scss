@import 'styles';

.selectInput {
  position: relative;

  > button {
    padding: $size-sm $size-sm;
    background: $white;
    width: 100%;
    cursor: pointer;

    .searchInputLabel {
      color: $gray-light;
      margin: 0.5rem 0;
    }

    .searchInputContent {
      display: flex;
      flex-direction: row;
      justify-content: space-between;
      align-items: center;
      border: 1px solid $gray-light;
      border-radius: 0.25rem;
      padding: 0.8rem 0.5rem;

      .selectArrowIcon {
        color: $gray-light;
      }
    }
  }

  .selectOptionList {
    position: absolute;
    bottom: -2.5rem;
    left: 1rem;
    right: 1rem;
    z-index: 50;
    box-shadow: $shadow-base;
    border-bottom-left-radius: 0.25rem;
    border-bottom-right-radius: 0.25rem;
    overflow: hidden;

    > li {
      > button {
        display: flex;
        align-items: center;
        justify-content: center;
        padding: 1rem 0;
        cursor: pointer;
        color: $gray-dark;
        z-index: 51;
        background: $white;
        width: 100%;
      }

      .selected {
        background: $orange;
        width: 100%;
      }
    }
  }
}
