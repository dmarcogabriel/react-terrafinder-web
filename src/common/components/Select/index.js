import React, { useState, useEffect } from 'react';
import { MdArrowDropUp, MdArrowDropDown } from 'react-icons/md';
import classes from './Select.module.scss';

export default function Select({
  label,
  value,
  options = [],
  id,
  onChange = () => {},
}) {
  const [showOptions, setShowOptions] = useState(false);
  const [defaultValue, setDefaultValue] = useState(value || 'Selecione...');
  const [optionsList, setOptionsList] = useState(options);

  const handleClickOutside = () => {
    // todo: handle click outside
    // const { classList } = e.target;
    // console.log(classList.contains('Select_option'));
    // if (!classList.contains('Select_option')) {
    //   setShowOptions(false);
    // }
  };

  const handleSelectOption = (e) => {
    setOptionsList(
      optionsList.map((option) => ({
        ...option,
        selected: option.key === e.key,
      }))
    );
    setDefaultValue(e.name);
    setShowOptions(false);
    onChange(e);
  };

  useEffect(() => {
    // document.addEventListener('mousedown', handleClickOutside);
    // return () => document.removeEventListener('mousedown', handleClickOutside);
  });

  return (
    <div className={classes.selectInput}>
      <button type="button" onClick={() => setShowOptions(!showOptions)}>
        <p className={classes.searchInputLabel} label={id}>
          {label}
        </p>

        <div className={classes.searchInputContent}>
          <p>{defaultValue}</p>

          {showOptions ? (
            <MdArrowDropUp className={classes.selectArrowIcon} size={22} />
          ) : (
            <MdArrowDropDown className={classes.selectArrowIcon} size={22} />
          )}
        </div>
      </button>

      {showOptions && (
        <ul className={classes.selectOptionList} id={id}>
          {optionsList.map((option) => (
            <li key={option.key}>
              <button
                type="button"
                className={option.selected && classes.selected}
                onClick={() => handleSelectOption(option)}
              >
                {option.name}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}
