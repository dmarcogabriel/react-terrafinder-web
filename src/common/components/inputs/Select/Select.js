import React from 'react';
import {
  Box,
  Select,
  InputLabel,
  MenuItem,
  FormHelperText,
} from '@mui/material';
import {
  Mood as ValidIcon,
  SentimentDissatisfied as InvalidIcon,
} from '@mui/icons-material';

const ValidationIcon = ({ errorMessage }) =>
  errorMessage ? (
    <InvalidIcon fontSize="small" color="error" />
  ) : (
    <ValidIcon fontSize="small" color="success" />
  );

export const SelectInput = ({
  value,
  label,
  onChange,
  dataTestId = 'select-input',
  options,
  variant = 'standard',
  noValidation,
  errorMessage,
  isLoading,
  inputProps = {},
  ...props
}) => {
  const handleChange = ({ target: { value: selectedOption } }) => {
    onChange(selectedOption);
  };

  return (
    <Box
      variant={variant}
      sx={{
        width: '100%',
        my: 2,
      }}
    >
      <Box
        sx={{
          display: 'flex',
          alingItems: 'center',
          justifyContent: 'space-between',
        }}
      >
        <InputLabel error={!!errorMessage}>{label}</InputLabel>
        {!noValidation && <ValidationIcon errorMessage={errorMessage} />}
      </Box>
      <Select
        inputProps={{ 'data-testid': dataTestId, ...inputProps }}
        value={isLoading ? 'Carregando...' : value}
        onChange={handleChange}
        fullWidth
        size="small"
        error={!!errorMessage}
        disabled={isLoading}
        {...props}
      >
        {!isLoading &&
          options.map((option, i) => (
            <MenuItem
              data-testid={`option-${i}`}
              key={String(i)}
              value={option}
            >
              {option}
            </MenuItem>
          ))}
      </Select>
      {!!errorMessage && (
        <FormHelperText sx={{ ml: '14px' }} error>
          {errorMessage}
        </FormHelperText>
      )}
    </Box>
  );
};
