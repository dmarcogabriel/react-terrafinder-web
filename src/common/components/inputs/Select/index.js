import { SelectInput } from './Select';

export { SelectInput };
