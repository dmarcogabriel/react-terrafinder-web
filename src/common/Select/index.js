import React, { useState, useEffect } from 'react';
import { MdArrowDropUp, MdArrowDropDown } from 'react-icons/md';
import { useTheme } from 'styled-components';
import cn from 'classnames';
import './styles.scss';

export default function Select({
  label,
  value,
  options = [],
  id,
  onChange = () => {},
}) {
  const [showOptions, setShowOptions] = useState(false);
  const [defaultValue, setDefaultValue] = useState(value || 'Selecione...');
  const [optionsList, setOptionsList] = useState(options);
  const { colors } = useTheme();

  const handleClickOutside = (e) => {
    const { classList } = e.target;

    if (
      !classList.contains('option') &&
      !classList.contains('selected-option')
    ) {
      setShowOptions(false);
    }
  };

  const handleSelectOption = (e) => {
    setOptionsList(
      optionsList.map((option) => {
        return option.key === e.key
          ? { ...option, selected: true }
          : { ...option, selected: false };
      })
    );
    setDefaultValue(e.name);
    setShowOptions(false);
    onChange(e);
  };

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);

    return () => document.removeEventListener('mousedown', handleClickOutside);
  });

  return (
    <div className="container">
      <button
        className="button"
        type="button"
        onClick={() => setShowOptions(!showOptions)}
      >
        <p label={id}>{label}</p>

        <div className="content">
          <p>{defaultValue}</p>

          {showOptions ? (
            <MdArrowDropUp style={{ color: colors.orange }} />
          ) : (
            <MdArrowDropDown style={{ color: colors.orange }} />
          )}
        </div>
      </button>

      {showOptions && (
        <ul className="optionsList" id={id}>
          {optionsList.map((option) => (
            <li>
              <button
                type="button"
                key={option.key}
                className={cn('option-button', option.selected && 'selected')}
                onClick={() => handleSelectOption(option)}
              >
                {option.name}
              </button>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
}
