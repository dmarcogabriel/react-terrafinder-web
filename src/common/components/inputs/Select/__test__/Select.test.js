import React from 'react';
import ReactDOM from 'react-dom';
// import { render, fireEvent, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { SelectInput } from '../Select';

const mockOptions = [
  {
    key: 't0',
    value: '',
    name: 'None',
  },
  {
    key: 't1',
    value: 'test1',
    name: 'Test 1',
  },
  {
    key: 't2',
    value: 'test2',
    name: 'Test 2',
  },
];

describe('common/components/input/Select', () => {
  it('should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <SelectInput label="test" options={mockOptions} value="" />,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should pass on click any option', async () => {
    // const { getByTestId, rerender } = render(
    //   <SelectInput label="test" options={mockOptions} />
    // );
    // const selectInput = getByTestId('select-input');
    // fireEvent.click(selectInput);
    // await waitFor(() =>
    //   rerender(<SelectInput label="test" options={mockOptions} />)
    // );
    // expect(getByTestId('option-t1')).toBeInTheDocument();
    // expect(getByTestId('option-t2')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    expect(
      renderer
        .create(
          <SelectInput
            options={mockOptions}
            label="Snapshot test"
            errorMessage="Error test message"
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
