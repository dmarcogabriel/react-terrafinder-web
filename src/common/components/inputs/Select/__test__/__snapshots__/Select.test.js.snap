// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/components/input/Select should match snapshot 1`] = `
<div
  className="MuiBox-root css-pcfq5i"
>
  <div
    className="MuiBox-root css-1663l76"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
    >
      Snapshot test
    </label>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeSmall css-1efkf0c-MuiSvgIcon-root"
      data-testid="SentimentDissatisfiedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <circle
        cx="15.5"
        cy="9.5"
        r="1.5"
      />
      <circle
        cx="8.5"
        cy="9.5"
        r="1.5"
      />
      <path
        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-3.5c.73 0 1.39.19 1.97.53.12-.14.86-.98 1.01-1.14-.85-.56-1.87-.89-2.98-.89-1.11 0-2.13.33-2.99.88.97 1.09.01.02 1.01 1.14.59-.33 1.25-.52 1.98-.52z"
      />
    </svg>
  </div>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-sizeSmall MuiSelect-root css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="notranslate"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      data-testid="select-input"
      onAnimationStart={[Function]}
      onChange={[Function]}
      tabIndex={-1}
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1a1ihhj"
      >
        <span
          className="notranslate"
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </legend>
    </fieldset>
  </div>
  <p
    className="MuiFormHelperText-root Mui-error css-qez4un-MuiFormHelperText-root"
  >
    Error test message
  </p>
</div>
`;
